packages/flame/lib/src/geometry/ray2.dart

```diff
@@ -0,0 +1,129 @@
+
+/// A ray in the 2d plane.
+///
+/// The [direction] should be normalized.
+class Ray2 {
+  Ray2(this.origin, Vector2 direction) {
+    this.direction = direction;
+  }
+
+  Ray2.zero() : this(Vector2.zero(), Vector2(1, 0));
+
+  Vector2 origin;
+  final Vector2 _direction = Vector2.zero();
+  Vector2 get direction => _direction;
+  set direction(Vector2 direction) {
+    assert(
+      (direction.length2 - 1).abs() < 0.000001,
+      'direction must be normalized',
+    );
+    _direction.setFrom(direction);
+    directionInvX = (1 / direction.x).toFinite();
+    directionInvY = (1 / direction.y).toFinite();
+  }
+
+  // These are the inverse of the direction (the normal), they are used to avoid
+  // a division in [intersectsWithAabb2], since a ray will usually be tried
+  // against many bounding boxes it's good to pre-calculate it, which is done
+  // in the direction setter.
+  @visibleForTesting
+  late double directionInvX;
+  @visibleForTesting
+  late double directionInvY;
+
+  /// Whether the ray intersects the [box] or not.
+  ///
+  /// Rays that originate on the edge of the [box] are considered to be
+  /// intersecting with the box no matter what direction they have.
+  // This uses the Branchless Ray/Bounding box intersection method by Tavian,
+  // but since +-infinity is replaced by +-maxFinite for directionInvX and
+  // directionInvY, rays that originate on an edge will always be considered to
+  // intersect with the aabb, no matter what direction they have.
+  // https://tavianator.com/2011/ray_box.html
+  // https://tavianator.com/2015/ray_box_nan.html
+  bool intersectsWithAabb2(Aabb2 box) {
+    final tx1 = (box.min.x - origin.x) * directionInvX;
+    final tx2 = (box.max.x - origin.x) * directionInvX;
+
+    final ty1 = (box.min.y - origin.y) * directionInvY;
+    final ty2 = (box.max.y - origin.y) * directionInvY;
+
+    final tMin = max(min(tx1, tx2), min(ty1, ty2));
+    final tMax = min(max(tx1, tx2), max(ty1, ty2));
+
+    return tMax >= max(tMin, 0);
+  }
+
+  /// Gives the point at a certain length along the ray.
+  Vector2 point(double length, {Vector2? out}) {
+    return ((out?..setFrom(origin)) ?? origin.clone())
+      ..addScaled(direction, length);
+  }
+
+  static final Vector2 _v1 = Vector2.zero();
+  static final Vector2 _v2 = Vector2.zero();
+  static final Vector2 _v3 = Vector2.zero();
+  static final Vector2 _v4 = Vector2.zero();
+
+  /// Returns where (length wise) on the ray that the ray intersects the
+  /// [LineSegment] or null if there is no intersection.
+  ///
+  /// A ray that is parallel and overlapping with the [segment] is considered to
+  /// not intersect.
+  double? lineSegmentIntersection(LineSegment segment) {
+    _v1
+      ..setFrom(origin)
+      ..sub(segment.from);
+    _v2
+      ..setFrom(segment.to)
+      ..sub(segment.from);
+    _v3.setValues(-direction.y, direction.x);
+
+    final dot = _v2.dot(_v3);
+
+    if (dot == 0) {
+      // ray is parallel to line
```

Comment:
Sorry I'm on the phone rn, but I'd do smth like this here:
1. Let dot2=v1.dot(v3). Then if dot2!=0 there's no intersection and we can return null.
2. Otherwise, let t1=direction.dot(segment.from-ray.origin) and t2=direction.dot(segment.to-ray.origin). (These can be expressed in terms of v1 and v2). Then if both t1 and t2 positive, return the min of them; if both negative return null; otherwise, return 0 because the Ray's origin is inside the segment.

This ensures that we return a point closest to Ray's origin, and correctly handle the case when ray and segment are on parallel lines.

Reply:
Hmm, maybe it makes sense that it doesn't collide with parallel rays, because it will always collide with one of the line segments connecting to it anyways? I'll remove it again for now and then we can re-add it again later if we choose to.